### PR TITLE
Editor interfaces

### PR DESCRIPTION
**Pull request by thebiglabasky**
Proposed basic support for editor interfaces.

### CODE REVIEW

File: src/lib/fetcher.ts [hunk 1]
```diff
@@ -1,18 +1,23 @@
 import IntentList from './intent-list'
-import APIContentType from '../lib/interfaces/content-type'
+import APIContentType, { APIEditorInterfaces } from '../lib/interfaces/content-type'
```
Comment (zcei) [inline src/lib/fetcher.ts]: `import { APIContentType, APIEditorInterfaces } from ...`

File: src/lib/fetcher.ts [hunk 2]
```diff
@@ -1,18 +1,23 @@
...
   constructor (makeRequest) {
     this.makeRequest = makeRequest
   }
 
-  async getEntriesInIntents (intentList: IntentList) {
-    const ids: string[] = _.uniq(intentList.getIntents().map((intent) => intent.getContentTypeId()))
+  async getEntriesInIntents (intentList: IntentList): Promise<Array<Object>> {
+    const ids: string[] = _.uniq(
+      intentList.getIntents()
+      .filter((intent) => intent.isContentTransform() || intent.isEntryDerive())
```
Comment (zcei) [inline src/lib/fetcher.ts]: Nice catch 👏

File: src/lib/fetcher.ts [hunk 3]
```diff
@@ -27,10 +32,13 @@ export default class Fetcher {
   }
 
   async getContentTypesInChunks (intentList: IntentList): Promise<APIContentType[]> {
-    const ids: string[] = _.uniq(intentList.getIntents().reduce((ids, intent) => {
-      const intentIds = intent.getRelatedContentTypeIds()
-      return ids.concat(intentIds)
-    }, []))
+    const ids: string[] = _.uniq(intentList.getIntents()
+      .filter((intent) => !intent.isEditorInterfaceUpdate())
```
Comment (zcei) [inline src/lib/fetcher.ts]: Per se this is correct, though I think it makes it harder to understand, as it implies EditInterface updates wouldn't need to know anything about ContentTypes, which isn't correct, right?
Given that the EditorInterface update happens on a ContentType, the ContentType will be fetched either way, so I would rather let the `_.uniq` sort out that duplication.

File: src/lib/fetcher.ts [hunk 4]
```diff
@@ -45,7 +53,47 @@ export default class Fetcher {
...
+
+    for (let id of contentTypeIds) {
+      try {
+        const response = await this.makeRequest({
+          method: 'GET',
+          url: `/content_types/${id}/editor_interface`
+        })
+        editorInterfaces.set(id, response)
+      } catch (error) {
+        if (error.name === 'NotFound') {
```
Comment (zcei) [inline src/lib/fetcher.ts]: We should think about exposing the status code, that might look a bit cleaner than checking against our own API error codes.

File: src/lib/fetcher.ts [hunk 5]
```diff
@@ -45,7 +53,47 @@ export default class Fetcher {
...
+
+    let editorInterfaces = new Map<string, APIEditorInterfaces>()
+
+    if (contentTypeIds.length === 0) {
+      return editorInterfaces
+    }
+
+    for (let id of contentTypeIds) {
+      try {
+        const response = await this.makeRequest({
```
Comment (zcei) [inline src/lib/fetcher.ts]: For readability we should move this to it's own method, as this is the only place where we currently allow resources to not exist, which makes the code a bit bloated.

But correct approach 👍

File: src/lib/fetcher.ts [hunk 6]
```diff
@@ -45,7 +53,47 @@ export default class Fetcher {
...
+          })
+        } else {
+          throw error
+        }
+      }
+    }
+    return editorInterfaces
+  }
+
+  async getLocalesForSpace (): Promise<Array<string>> {
```
Comment (zcei) [inline src/lib/fetcher.ts]: We're trying to get rid of the `Array` generic, so please:
`Promise<string[]>`

File: src/lib/intent/base-intent.ts [hunk 7]
```diff
@@ -115,6 +121,20 @@ export default abstract class Intent implements IntentInterface {
...
+   * @returns The API action to perform on save. Defaults to saving the Content Type.
+   */
+  getSaveAction (): APIAction {
+    return new ContentTypeSaveAction(this.getContentTypeId())
+  }
+
+  /**
+   * @returns The API action to perform on publish. Defaults to publishing the Content Type.
+   */
+  getPublishAction (): APIAction {
```
Comment (zcei) [inline src/lib/intent/base-intent.ts]: This block doesn't seem to be used at all?
Also you can just import the `ContentTypeSaveAction` / `ContentTypePublishAction` in your intent and add it as an action. See `src/lib/intent/field-delete.ts` for an example.

File: src/lib/intent/composed-intent.ts [hunk 8]
```diff
@@ -36,6 +37,15 @@ export default class ComposedIntent implements Intent {
   }
 
   // TODO: adjust interface so we don't have to implement all this
+  getSaveAction (): APIAction {
+    return null
+  }
+  getPublishAction (): APIAction {
```
Comment (zcei) [inline src/lib/intent/composed-intent.ts]: This just mirrors the interface from above and therefore isn't used as well.

File: src/lib/intent/editorinterface-update.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    return true
+  }
+  shouldSave (): boolean {
+    return false
+  }
+  shouldPublish (): boolean {
+    return false
+  }
+  toActions () {
+    // TODO: check for optimizing API calls and saving CT by CT with merged editor interface definition
```
Comment (zcei) [inline src/lib/intent/editorinterface-update.ts]: Is this possible? It is its on API call, never seen it merged into a ContentType update

File: src/lib/offline-api/index.ts [hunk 10]
```diff
@@ -1,7 +1,7 @@
 import { omit, compact } from 'lodash'
 import FieldDeletionValidator from './validator/field-deletion'
 import { ContentTypePayloadValidator } from './validator/content-type'
-import ContentType from '../entities/content-type'
+import ContentType, { EditorInterfaces } from '../entities/content-type'
```
Comment (zcei) [inline src/lib/offline-api/index.ts]: `import { ContentType, EditorInterfaces } from ...`

File: src/lib/intent/composed-intent.ts (hunk 8 above)
Comment (thebiglabasky) [reply src/lib/intent/composed-intent.ts]: Ah yes, remains from an attempt to make the save content type and publish content type actions generic.
I'll get rid of these

File: src/lib/intent/editorinterface-update.ts (hunk 9 above)
Comment (thebiglabasky) [reply src/lib/intent/editorinterface-update.ts]: Nah, I wanted ideally to merge editor interface changes of the same content type in a single API call, but that doesn't seem to be easy to do.
Maybe as a follow up!

File: src/lib/fetcher.ts (hunk 3 above)
Comment (thebiglabasky) [reply src/lib/fetcher.ts]: In theory, you don't need to fetch the content type if the only thing you do is updating the editor interface though.
I'd suggest we keep, and add a comment?

File: src/lib/fetcher.ts (hunk 4 above)
Comment (thebiglabasky) [reply src/lib/fetcher.ts]: I'll add a TODO

File: src/lib/fetcher.ts (hunk 3 above)
Comment (zcei) [reply src/lib/fetcher.ts]: Ah you're right, the `editContentType` doesn't dispatch an intent.

**Comment by thebiglabasky** [comment]
@zcei Could you PTAL? Thanks!

**Pull request merged**